controllers/storagecluster/server.go

```diff
@@ -0,0 +1,187 @@
+
+	var finalErr error
+
+	for _, obj := range []client.Object{
+		desiredServerDeployment,
+		desiredServerService,
+	} {
+		err := r.Client.Delete(context.TODO(), obj)
+
+		if err != nil && !errors.IsNotFound(err) {
+			r.Log.Error(err, "nigoyal Failed to delete resource", "Kind", obj.GetObjectKind(), "Name", obj.GetName())
+			multierr.AppendInto(&finalErr, err)
+		}
+	}
+
+	if finalErr == nil {
+		r.Log.Info("nigoyal Resource deletion for provider succeeded")
+	}
+
+	return finalErr
+}
+
+func (obj *ocsProviderServer) isDeploymentReady(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	desiredServerDeployment := GetProviderServerDeployment(instance)
+	actualServerDeployment := &appsv1.Deployment{}
+
+	err := r.Client.Get(context.TODO(), types.NamespacedName{
+		Name:      desiredServerDeployment.Name,
+		Namespace: desiredServerDeployment.Namespace,
+	}, actualServerDeployment)
+
+	if err != nil {
+		r.Log.Error(err, "nigoyal Failed to get deployment", "Name", desiredServerDeployment.Name)
+		return err
+	}
+
+	if actualServerDeployment.Status.AvailableReplicas != *desiredServerDeployment.Spec.Replicas {
+		r.Log.Error(err, "nigoyal Deployment is not ready", "Name", desiredServerDeployment.Name,
+			"AvailableReplicas", actualServerDeployment.Status.AvailableReplicas)
+		return fmt.Errorf("Deployment %s is not ready", desiredServerDeployment.Name)
+	}
+
+	r.Log.Info("nigoyal Deployment is working as expected", "Name", desiredServerDeployment.Name)
+	return nil
+}
+
+func GetProviderServerDeployment(instance *ocsv1.StorageCluster) *appsv1.Deployment {
+
+	return &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      generateProviderDeploymentName(),
+			Namespace: instance.Namespace,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: func(i int32) *int32 { return &i }(1),
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app": "apiServer",
+				},
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app": "apiServer",
+					},
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:    "server",
+							Image:   ServerImage,
+							Command: []string{"/usr/local/bin/consumer-api"},
+							Ports: []corev1.ContainerPort{
+								{
+									Name:          "api-server-port",
+									ContainerPort: int32(50051),
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func GetProviderServerService(instance *ocsv1.StorageCluster) *corev1.Service {
+
+	return &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      generateProviderDeploymentName(),
+			Namespace: instance.Namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Selector: map[string]string{
+				"app": "apiServer",
+			},
+			Ports: []corev1.ServicePort{
+				{
+					Port:       int32(50051),
+					TargetPort: intstr.FromString("api-server-port"),
+					NodePort:   int32(30080),
```

Comment:
We can use node ports in range b/w `30000-32767`, Do we need any kind of approval before using any of the port?

Reply:
I am not sure about that I can give it a try. @sp98 any idea?

Reply:
This is not required: https://v1-19.docs.kubernetes.io/docs/concepts/services-networking/service/#nodeport
```
      # Optional field
      # By default and for convenience, the Kubernetes control plane will allocate a port from a range (default: 30000-32767)
      nodePort: 30007
```

Reply:
removed